rust_src/src/eval.rs

```diff
@@ -1338,4 +1338,132 @@ pub fn condition_case(args: LispCons) -> LispObject {
+/// can be referred to by the Emacs help facilities and other programming
+/// tools.  The `defvar' form also declares the variable as \"special\",
+/// so that it is always dynamically bound even if `lexical-binding' is t.
+///
+/// If SYMBOL's value is void and the optional argument INITVALUE is
+/// provided, INITVALUE is evaluated and the result used to set SYMBOL's
+/// value.  If SYMBOL is buffer-local, its default value is what is set;
+/// buffer-local values are not affected.  If INITVALUE is missing,
+/// SYMBOL's value is not set.
+///
+/// If SYMBOL has a local binding, then this form affects the local
+/// binding.  This is usually not what you want.  Thus, if you need to
+/// load a file defining variables, with this form or with `defconst' or
+/// `defcustom', you should always load that file _outside_ any bindings
+/// for these variables.  (`defconst' and `defcustom' behave similarly in
+/// this respect.)
+///
+/// The optional argument DOCSTRING is a documentation string for the
+/// variable.
+///
+/// To define a user option, use `defcustom' instead of `defvar'.
+/// usage: (defvar SYMBOL &optional INITVALUE DOCSTRING)
+#[lisp_fn(min = "1", unevalled = "true")]
+pub fn defvar(args: LispCons) -> LispObject {
+    let (sym_obj, tail) = args.into();
+
+    if tail.is_not_nil() {
+        if LispCons::from(tail).length() > 2 {
+            error!("Too many arguments");
+        }
+
+        let sym: LispSymbolRef = sym_obj.into();
+        let tem = default_boundp(sym);
+
+        // Do it before evaluating the initial value, for self-references.
+        sym.set_declared_special(true);
+
+        if tem {
+            // Check if there is really a global binding rather than just a let
+            // binding that shadows the global unboundness of the var.
+            let mut binding = default_toplevel_binding_rust(sym);
+            if !binding.is_null() && (binding.specpdl_old_value() == Qunbound) {
+                binding.set_specpdl_old_value(unsafe { eval_sub(car(tail)) });
+            }
+        } else {
+            set_default(sym, unsafe { eval_sub(car(tail)) });
+        }
+
+        let mut tem = car(cdr(tail));
+
+        if tem.is_not_nil() {
+            if unsafe { globals.Vpurify_flag }.is_not_nil() {
+                tem = purecopy(tem);
+            }
+            put(sym, Qvariable_documentation, tem);
+        }
+        loadhist_attach(sym_obj);
+    } else if unsafe { globals.Vinternal_interpreter_environment }.is_not_nil()
+        && sym_obj
+            .as_symbol()
+            .map(|x| !x.get_declared_special())
+            .unwrap_or(false)
+    {
+        // A simple (defvar foo) with lexical scoping does "nothing" except
+        // declare that var to be dynamically scoped *locally* (i.e. within
+        // the current file or let-block).
+        unsafe {
+            globals.Vinternal_interpreter_environment =
+                Fcons(sym_obj, globals.Vinternal_interpreter_environment);
```

Comment:
The preferred way of writing this is `LispObject::cons(sym_obj, globals.Vinternal_interpreter_environment)`

Reply:
Done